### PR TITLE
Support Devfile endpoint annotations in Che Router

### PR DESCRIPTION
**Pull request by AObuchow**
### What does this PR do?
- Adds the DWR endpoint annotations as annotations for the route (on OpenShift) or ingresse (on Kubernetes) created by the Che Router for a given endpoint

There are a few additional changes made in this PR that are not meant to be merged, though they may be desirable:
- Using per-workspace storage as well as the UDI in the Che-Operator devfile. This was helpful for developing this patch in Che.

### Screenshot/screencast of this PR
n/a

### What issues does this PR fix or reference?
Fix eclipse-che/che#23064

### How to test this PR?
1. Install DWO built from this [PR](https://github.com/devfile/devworkspace-operator/pull/1297), as well as Che-Operator built from the current PR onto your cluster. 
  - On OpenShift, this can easily be done by running `./build/scripts/olm/test-catalog-from-sources.sh ` from the root of this repo.
  - On minikube/kubernetes, you'll have to:
    1. Build DWO: From the root of the [DWO repo](https://github.com/devfile/devworkspace-operator) run `export DWO_IMG=quay.io/<username>/devworkspace-controller:endpoint-annotations && make docker` (or just use my image:  `export DWO_IMG=quay.io/<username>/devworkspace-controller:endpoint-annotations-implementation `)
    2. Install DWO: `make install_cert_manager && make install`
    3. Build Che-Operator: `make docker-build docker-push IMG=<YOUR_OPERATOR_IMAGE>`
    4. Install Che: `chectl server:deploy -p minikube --che-operator-image=<YOUR_OPERATOR_IMAGE>`

2. Once both DWO and Che are installed onto your cluster, verify there's a Che Cluster CR instance created so that you can access the Che User dashboard
3. From the Che User Dashboard, create a workspace with a devfile that defines endpoints with annotations, such as [this one](https://github.com/AObuchow/devfile-endpoint-annotation-test/blob/main/devfile.yaml).

```YAML
      endpoints:
        - name: https-python
          targetPort: 8080
          protocol: https
          annotation:
            my-annotation: test
            another-annotation: test2
```

4. Once the workspace has started up, verify the routes & ingresses for the endpoint(s) with annotations have the expected annotations from the devfile. In my example devfile, your `<workspaceid>-py-8080-https-python` route/ingress should have `my-annotation: test` & `another-annotation: test2`, as shown below:

```diff
kind: Route
apiVersion: route.openshift.io/v1
metadata:
  annotations:
+    another-annotation: test2
    che.routing.controller.devfile.io/component-name: py
    che.routing.controller.devfile.io/endpoint-name: https-python
+    my-annotation: test
  resourceVersion: '48291'
  name: workspace4b929cbc569a4473-py-8080-https-python
(...)
  namespace: che-kube-admin-che-rwxo67
  ownerReferences:
    - apiVersion: controller.devfile.io/v1alpha1
      kind: DevWorkspaceRouting
      name: routing-workspace4b929cbc569a4473
      uid: 497a6df3-9e14-47f8-9fc6-33acfda6df77
      controller: true
      blockOwnerDeletion: true
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    controller.devfile.io/devworkspace_id: workspace4b929cbc569a4473
spec:
(...)
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Development resource are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-development-resources)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: controllers/devworkspace/solver/endpoint_exposer.go
Comment (AObuchow) [inline controllers/devworkspace/solver/endpoint_exposer.go:259]: **Note:** the devfile endpoint annotations can actually override the values of the [Che Cluster CR ingress annotations](https://github.com/eclipse-che/che-operator/blob/main/api/v2/checluster_types.go#L238-L245). To me, this makes sense, as it gives users more fine-grained control over the ingress annotations. However, I don't have a strong opinion on this.

**Comment by AObuchow** [comment]
@tolusha Though this PR is still a draft, it should be good for review if you want to take a look already. In order for this PR to no longer be a draft, we'll need to release DWO 0.31.0, and I'll update this PR to pull in the latest version of DWO

File: controllers/devworkspace/solver/endpoint_exposer.go
Comment (AObuchow) [reply controllers/devworkspace/solver/endpoint_exposer.go:259]: I realized we are not adding annotations from the Che Cluster CR's `networking.annotations` fields to the routes created by the Che Router, only the ingresses get these annotations.

This seems like a bug or an error in the documentation, so I've opened https://github.com/eclipse-che/che/issues/23118 to track this. If this is indeed a bug, I could add the fix as part of this PR since it's closely related.

**Comment by tolusha** [comment]
Sounds good for me

**Comment by AObuchow** [comment]
@tolusha the DWO 0.31.0 release is currently in progress. Once it's released, I will update this PR and it will be ready for review :)

**Comment by AObuchow** [comment]
@tolusha this PR is finally ready for review now that DWO 0.31.0 is [being used](https://github.com/eclipse-che/che-operator/pull/1899/commits/511b5b7f845428ef0953d7d5852dc80f57471ad6) in Che-Operator :)

**Comment by AObuchow** [comment]
@tolusha by the way, if you end up wanting the devfile changes, I'll rewrite the commit messages for them (since they're currently marked as "do not merge").

**Comment by AObuchow** [comment]
Not sure why the pro v14-che-behind-proxy test is failing, though it seems unrelated to this PR:

```
INFO[2024-09-19T17:04:27Z] Logs for container test in pod che-behind-proxy-ipi-conf-aws-blackholenetwork: 
INFO[2024-09-19T17:04:27Z] 
An error occurred (AlreadyExistsException) when calling the CreateStack operation: Stack [ci-op-mfqq3b2j-9dadd-shared-vpc-blackhole] already exists
{"component":"entrypoint","error":"wrapped process failed: exit status 254","file":"sigs.k8s.io/prow/pkg/entrypoint/run.go:84","func":"sigs.k8s.io/prow/pkg/entrypoint.Options.internalRun","level":"error","msg":"Error executing test process","severity":"error","time":"2024-09-19T17:04:27Z"}
error: failed to execute wrapped command: exit status 254 
INFO[2024-09-19T17:04:27Z] Step che-behind-proxy-ipi-conf-aws-blackholenetwork failed after 9s. 
(...)
INFO[2024-09-19T17:06:58Z] Ran for 4m36s                                
ERRO[2024-09-19T17:06:58Z] Some steps failed:                           
ERRO[2024-09-19T17:06:58Z] 
  * could not run steps: step che-behind-proxy failed: "che-behind-proxy" pre steps failed: "che-behind-proxy" pod "che-behind-proxy-ipi-conf-aws-blackholenetwork" failed: could not watch pod: the pod ci-op-mfqq3b2j/che-behind-proxy-ipi-conf-aws-blackholenetwork failed after 7s (failed containers: test): ContainerFailed one or more containers exited
Container test exited with code 254, reason Error
---
An error occurred (AlreadyExistsException) when calling the CreateStack operation: Stack [ci-op-mfqq3b2j-9dadd-shared-vpc-blackhole] already exists
{"component":"entrypoint","error":"wrapped process failed: exit status 254","file":"sigs.k8s.io/prow/pkg/entrypoint/run.go:84","func":"sigs.k8s.io/prow/pkg/entrypoint.Options.internalRun","level":"error","msg":"Error executing test process","severity":"error","time":"2024-09-19T17:04:27Z"}
error: failed to execute wrapped command: exit status 254
--- 
```

**Review by ibuziuk — APPROVED** [review]

File: devfile.yaml
Comment (tolusha) [inline devfile.yaml:20]: Please, revert this line.
dev image is based on quay.io/devfile/universal-developer-image:ubi8-latest

File: controllers/devworkspace/solver/endpoint_exposer.go
Comment (tolusha) [inline controllers/devworkspace/solver/endpoint_exposer.go:253]: Is the order of parameters ingressAnnotations and endpointAnnotations correct?

Comment (AObuchow) [reply controllers/devworkspace/solver/endpoint_exposer.go:253]: Yes: `ingressAnnotations` refers to the annotations to add to the [ingress provided by the Che Cluster CR](https://github.com/eclipse-che/che-operator/blob/aa0150d74fe0db59f475192ba13230c68dfd7e72/api/v2/checluster_types.go#L244-L251), and `endpointAnnotatiosn` refers to the annotations provided by the devfile. It's worth ensuring we want the devfile annotations to be able to override the Che Cluster CR annotations, see https://github.com/eclipse-che/che-operator/pull/1881/files#r1714630847

When [we call this function](https://github.com/eclipse-che/che-operator/pull/1881/files#diff-d071c0e02e16bcc2c5c74251891232a82f99aa56ae415f7b4e12178e28bd7dccR201), we're first providing the IngressExposer's annotations (coming from the Che Cluster CR, see the last parameter of [this function call](https://github.com/eclipse-che/che-operator/blob/aa0150d74fe0db59f475192ba13230c68dfd7e72/controllers/devworkspace/solver/che_routing.go#L457)) and then the endpoint's annotations (coming from the devfile through the[ devworkspace routing's endpoint](https://github.com/eclipse-che/che-operator/pull/1881/files#diff-ebca2eefe12f7ba4a722c53d574ba1b2adee412909da8cdbc974c8f7fcbfb02fR517))
```go
Annotations:     finalizeIngressAnnotations(endpoint.annotations, e.ingressAnnotations, endpoint.componentName, endpoint.endpointName),
```

Hopefully this clarifies things - let me know if I misinterpreted your question.

File: devfile.yaml
Comment (AObuchow) [reply devfile.yaml]: Done, I removed the associated commit. However, the reason why I changed the image being used is because the che-operator-dev container image is outdated and needs to be rebuilt to pull in the latest changes from the UDI.

File: controllers/devworkspace/solver/endpoint_exposer.go
Comment (dkwon17) [reply controllers/devworkspace/solver/endpoint_exposer.go:253]: In line 201, should it be the other way around, like this?
```
Annotations:     finalizeIngressAnnotations(e.ingressAnnotations, endpoint.annotations, endpoint.componentName, endpoint.endpointName),
```

Comment (AObuchow) [reply controllers/devworkspace/solver/endpoint_exposer.go:253]: Ah I see now, can't believe I made that mistake. Good catch @tolusha @dkwon17 I'll fix this immediately 👍

Comment (AObuchow) [reply controllers/devworkspace/solver/endpoint_exposer.go:253]: Should be good now, see https://github.com/eclipse-che/che-operator/pull/1881/commits/d6c17245142d036418928b63ab4fb8ab4d8cd168.

Once verified and approved, I'll squash the fixup commit :)

**Review by dkwon17 — APPROVED** [review]

**Pull request merged**
